types/types.go

```diff
@@ -209,3 +183,29 @@ type Metrics struct {
 	// Cumulative size of all elements in memory cache (in bytes)
 	SizeMemoryCache uint64
 }
+
+// Array is a wrapper around a slice that ensures that we get "[]" JSON for nil values
+type Array[C any] []C
+
+// MarshalJSON ensures that we get "[]" for nil arrays
+func (a Array[C]) MarshalJSON() ([]byte, error) {
+	if len(a) == 0 {
+		return []byte("[]"), nil
+	}
+	var raw []C = a
+	return json.Marshal(raw)
+}
+
+// UnmarshalJSON ensures that we get nil for "[]" and "null"
+func (a *Array[C]) UnmarshalJSON(data []byte) error {
+	// make sure we deserialize [] back to null
+	if string(data) == "[]" || string(data) == "null" {
+		return nil
+	}
+	var raw []C
+	if err := json.Unmarshal(data, &raw); err != nil {
+		return err
+	}
+	*a = raw
+	return nil
```

Comment:
I know it was like this before. But to we really prefer nil over an empty slice? I guess we used nil in the past since `[]foo` defaults to nil and default values are kindof common in Go.

Reply:
I was wondering about that too. I think it makes sense to normalize to one of the two, so the callers don't have to check for both empty slice and nil all the time. Nil checks are also pretty common already.
One thing we lose with this is the distinction between the contract sending `null` or `[]`, but at least with a Rust contract that only matters for `Option<Vec<T>>` fields.

Reply:
> One thing we lose with this is the distinction between the contract sending null or [], but at least with a Rust contract that only matters for Option<Vec<T>> fields.

Yeah that's pretty much the goal. Go slices enforce having the two states but we only want `Vec<T>` in the vast majority of cases

Reply:
Well yes. Looking at it from the standpoint of the different states, it might make more sense to use an empty slice, since it more accurately represents what the Rust side sent, but I think the ergonomics on the Go side are better if we use `nil`.